---
title: "Differential Gene Expression Analysis Demonstration"
author: "William Rieger, Florian Herzler, Maximilian Otto"
date: "`r Sys.Date()`"

output:
  reprex::reprex_document:
    venue: "gh"
    advertise: FALSE
    session_info: TRUE
    style: TRUE
    comment: "#;-)"
    tidyverse_quiet: FALSE
    std_out_err: TRUE
knit: reprex::reprex_render
---

# DEseq2 - Background

DESeq2 is a widely used statistical tool for analyzing high-throughput RNA sequencing (RNA-seq) data. It is specifically designed for differential gene expression analysis, which involves comparing gene expression levels between different conditions or experimental groups.

### What it does

The main goal of DESeq2 is to identify genes that show statistically significant differences in expression between different conditions or groups. It calculates a measure called the "log-fold change" that quantifies the magnitude of change in gene expression between conditions, and a "p-value" that indicates the statistical significance of the observed differences.

DESeq2 uses a **negative binomial distribution model** to account for the inherent variability in RNA-seq data, where the mean and variance of gene expression are linked. It performs data normalization, estimation of size factors to account for differences in library sizes, and estimation of dispersion to capture the biological and technical variability.

### DEseq2 works out of the box, and hides many of its decisions.

Other DE packages such as LIMMA require you to manually employ multiple testing correction methods and preprocessing steps such as normalisation. DESeq2 takes care of these steps and can be tuned to your liking using its many parameters.

DESeq2 also employs multiple testing correction methods, such as the Benjamini-Hochberg procedure, to control for false discoveries resulting from testing a large number of genes simultaneously. This ensures that only genes with a low probability of being falsely identified as differentially expressed are reported.

Overall, DESeq2 provides a comprehensive and robust framework for differential gene expression analysis, enabling researchers to identify genes that are likely to play a role in biological processes or disease conditions based on RNA-seq data.

---

# Demonstration project

### ***Setup***

To install the packages below, please uncomment the bottom lines of code.

These packages often have installation problems... such is programming with R.

```{r pacman}
# install.packages("pacman")

# Installs packages if not already installed then loads them.
pacman::p_load("BiocManager", "DESeq2", "biomaRt", "clusterProfiler", "org.Hs.eg.db", "topGO", "Rgraphviz", "IHW", "tidyverse" )

```

In case Pacman doesn't work:

```{r Install, message=FALSE}
if (FALSE) {
  if (!require("BiocManager", quietly = TRUE))
      install.packages("BiocManager")

  # BiocManager::install("DEseq2")
  # BiocManager::install("GEOquery")
  # BiocManager::install("biomaRt")
  # BiocManager::install("clusterProfiler")
  # BiocManager::install("org.Hs.eg.db")
  # BiocManager::install("topGO")
  # BiocManager::install("Rgraphviz")
  # BiocManager::install("IHW")

  library(tidyverse)
  library(dplyr)
  library(DESeq2)
  library(GEOquery)
  library(biomaRt)
  library(clusterProfiler)
}
```

Load the packages. If other packages are not installed then install them quickly in the console.


### ***Data***

For the purpose of this exercise, we are going to use a publically available dataset from Gene Expression Omnibus (GEO).

GEO Dataset:
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE213691

The processed data was downloaded in txt format so we must read it in using the read.delim package. Seeing as the data is already in xlsx format, feel free to explore the data using excel.

```{r DataImport}
#load the count data
counts <- read.delim("/Users/matanatmammadli/Desktop/R Codes/GSE213691_CountTableHuman.txt", row.names=1)
meta <- as.data.frame(c("control", "PGE2", "TL", "PGE2TL",
                        "control", "PGE2", "TL", "PGE2TL",
                        "control", "PGE2", "TL", "PGE2TL",
                        "control", "PGE2", "TL", "PGE2TL"))
rownames(meta) <- colnames(counts)
colnames(meta) <- c("treatment")
```

**Data:** 

Expression profiling by high throughput sequencing (counts)

**Summary:**
RNA sequencing was used to characterize PGE2-mediated changes in the gene expression profile of human conventional type 1 dendritic cells (cDC1) purified from PBMCs of healthy donors. Our analysis shows that treatment of cDC1 with PGE2 induces transcriptional changes in resting cDC1. cDC1 activated with a TLR3 ligand after PGE2 pre-treatment show alterations in the expression of activation induced genes.

*PGE2 is a drug, involved in many different clinical situations, such as induction cantraction when giving birth, or changing cDC1 activity for different T-cell responses.*

### ***Experimental design:***

cDC1 (CD370+CD141+CD3- cells) were sort purified from PBMCs of four healthy donors (A,B,C,D). cDC1 were treated or not with PGE2 (1Âµg/ml) for 24h. In some groups, the TLR3 ligand poly I:C was added to the culture for the last 4h of the experiment to induce cDC1 activation.

---

### DESeq2 Analysis

First we want to create the DESeq2 dataset object.

The DESeq() function essentially does all of the work for you, and most of the time, the default settings will suffice.

Please refer to the documentation of this function for more information: 
https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/DESeq#:~:text=DESeq%3A%20Differential%20expression%20analysis%20based,(a.k.a.%20Gamma%2DPoisson)%20distribution

**Which test should I use?**

DESeq2 offers two kinds of hypothesis tests: the Wald test, where we use the estimated standard error of a log2 fold change to test if it is equal to zero, and the likelihood ratio test (LRT). The LRT examines two models for the counts, a full model with a certain number of terms and a reduced model, in which some of the terms of the full model are removed. The test determines if the increased likelihood of the data using the extra terms in the full model is more than expected if those extra terms are truly zero.

The LRT is therefore useful for testing multiple terms at once, for example testing 3 or more levels of a factor at once, or all interactions between two variables.

The LRT test is particularly useful for time series tests.

```{r DESeq2Run}
# Filtering out the control and dual treatment, probably a nicer way.
counts2 <- counts[,c(1,4,5,8,9,12,13,16)]
meta2 <- as.data.frame(meta[c(1,4,5,8,9,12,13,16),])
rownames(meta2) <- colnames(counts2)
colnames(meta2) <- c("treatment")


dds <- DESeqDataSetFromMatrix(countData = as.matrix(counts2),
                              colData = as.matrix(meta2),
                              design = ~ treatment)
dds <- DESeq(dds, test="Wald")
```

By default, DESeq2 normalizes the counts within each sample, allowing for more reliable and meaningful comparisons of gene expression levels across samples.

DESeq2 employs the size factors normalization method as the default approach for normalizing gene expression data. The size factors estimation takes into account the library size differences between samples to ensure accurate comparisons.

### **Task:** 

Explore and sort the results `res` so that the most significantly differentially expressed genes appear first in our dataframe.

**Question:** Why must we adjust our p-values? Write a 2 sentence answer below.

**Question:** How many differentially expressed genes are there?

```{r NumberOFDEGs}
res <- results(dds)
view(res)
```

DESeq2 usually applies the Benjamini-Hochberg (BH) method as the default procedure for multiple testing correction. 

A generalization of the idea of p value filtering is to weight hypotheses to optimize power. A Bioconductor package, IHW, is available that implements the method of Independent Hypothesis Weighting (Ignatiadis et al. 2016). Here we show the use of IHW for p value adjustment of DESeq2 results. For more details, please see the vignette of the IHW package. The IHW result object is stored in the metadata.

DESeq2 provides flexibility in setting the p-value threshold for filtering. Researchers can choose the threshold based on their specific study objectives, statistical rigor, and considerations of multiple testing corrections to control the false discovery rate (FDR) using methods such as the Benjamini-Hochberg procedure.

**Question:** Can you think of scenarios or reasons one might be justified in filtering with a p-value threshold greater than 0.05? Write your answer in dot-points.

```{r IHW}
##install.packages("IHW")
library(IHW)
resIHW <- results(dds, filterFun=ihw)
summary(resIHW)
sum(resIHW$padj < 0.1, na.rm=TRUE)
# plot(metadata(resIHW)$ihwResult)
```



## Assessing the quality of our model/results

```{r DispEsts}
plotDispEsts(dds, cex = 1)
```

The dispersion plot in DESeq2 provides valuable insights into the variability or dispersion of gene expression across different expression levels. It helps assess the goodness-of-fit of the model used by DESeq2 to estimate differential expression.

In the dispersion plot, each point represents a gene, and the x-axis typically shows the mean expression level of the gene, while the y-axis represents the estimated dispersion. The dispersion indicates the degree of variation in gene expression that cannot be accounted for by the model's assumptions.

**Question:** What can the dispersion plot tell us about our results? Write a short 2 sentence answer.

```{r PlotMA}
plotMA(res, ylim = c(-8, 8), cex = 1)
```


### ***Getting Better Gene Annotations***

The database didn't report the data in the regular format with proper gene annotations as is usually* found in GEO.

*Note: GEO database is actually VERY messy.

```{r annotatingGenes}
# library(biomaRt)

ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", mirror = "www")

ids <- rownames(res)

annotations <- getBM(attributes = c("ensembl_gene_id", 'entrezgene_id', 'external_gene_name'),
                    filters = "ensembl_gene_id",
                    values = ids,
                    mart = ensembl)


view(annotations)

```

```{r}
view(res)
```

### **Task:** 

Now explore the results and use online databases to find an interesting gene that appears to be up or down regulated. Give your biological interpretation for why it is regulated in such a way.

### **GO-term Enrichment Analysis**

```{r gotermAnaly}
# library(org.Hs.eg.db)

de_genes <- rownames(subset(res, padj < 0.05 & abs(log2FoldChange) > 1))

go_results <- enrichGO(gene          = de_genes,
                        keyType       = "ENSEMBL",
                        OrgDb         = org.Hs.eg.db,
                        ont           = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05,
                        readable      = TRUE)

```

### **Visualising the GO-term Enrichment Analysis**

The four following plots utilize different visualization techniques to represent gene ontology analysis results, providing insights into the distribution of gene associations, relationships between categories, and the hierarchical structure of the gene ontology.

```{r Dotplot, fig.height=10, fig.width=10}
dotplot(go_results, showCategory=30)
```

The dot plot represents the categories or terms from the GO analysis on the y-axis, and the number of genes associated with each category on the x-axis. The showCategory parameter specifies the number of top categories to display in the plot. This plot is useful for gaining an overview of the distribution of gene associations across different GO categories.

```{r CNET, fig.height=10, fig.width=10}
cnetplot(go_results, foldChange=de_genes)
```

The network plot represents the GO categories as nodes, and the edges between the nodes indicate relationships between the categories based on the gene associations. This plot helps identify clusters or groups of related GO categories.

```{r Circular-CNET, fig.height = 10, fig.width = 10}
cnetplot(go_results, foldChange = de_genes, circular = TRUE, colorEdge = TRUE)
```

By setting circular to TRUE, the plot is arranged in a circular layout, which can provide a different perspective on the relationships. The colorEdge parameter set to TRUE colors the edges connecting the nodes in the network, making it easier to distinguish between different relationships.

```{r GO-Graph, fig.height=10, fig.width=10}
plotGOgraph(go_results)
```

The graph represents the GO categories as nodes and displays the relationships between them using edges. The plot provides a more detailed view of the hierarchical structure of the GO categories and their connections. This visualization is useful for exploring the relationships between specific categories and understanding the overall structure of the gene ontology analysis results.

```{r meme}
# Open YouTube video in the default browser
browseURL("https://www.youtube.com/watch?v=sIlNIVXpIns")
```